Apache httpclient unbundled with 223.7126.7

```
java.lang.NoClassDefFoundError: org/apache/commons/httpclient/HttpConnectionManager
	at org.codinjutsu.tools.jenkins.security.SecurityClientFactory.basic(SecurityClientFactory.java:35)
	at org.codinjutsu.tools.jenkins.logic.RequestManager.authenticate(RequestManager.java:295)
	at org.codinjutsu.tools.jenkins.logic.LoginService$1.run(LoginService.java:42)
	at org.codinjutsu.tools.jenkins.logic.JenkinsBackgroundTask.run(JenkinsBackgroundTask.java:41)
	at com.intellij.openapi.progress.impl.CoreProgressManager.startTask(CoreProgressManager.java:425)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.startTask(ProgressManagerImpl.java:114)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcessWithProgressAsynchronously$6(CoreProgressManager.java:476)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$3(ProgressRunner.java:252)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:190)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$executeProcessUnderProgress$13(CoreProgressManager.java:591)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:666)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computeUnderProgress(CoreProgressManager.java:622)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:590)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:60)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:177)
	at com.intellij.openapi.progress.impl.ProgressRunner.lambda$submit$4(ProgressRunner.java:252)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:702)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1$1.run(Executors.java:699)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.util.concurrent.Executors$PrivilegedThreadFactory$1.run(Executors.java:699)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.ClassNotFoundException: org.apache.commons.httpclient.HttpConnectionManager PluginClassLoader(plugin=PluginDescriptor(name=Jenkins Control, id=Jenkins Control Plugin, descriptorPath=plugin.xml, path=P:\JetBrains\Toolbox\apps\IDEA-U\ch-0\223.7126.7.plugins\jenkins-control-plugin, version=0.13.14-2022.2, package=null, isBundled=false), packagePrefix=null, instanceId=168, state=active)
	at com.intellij.ide.plugins.cl.PluginClassLoader.loadClass(PluginClassLoader.java:215)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	... 24 more


```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,6 +1,8 @@
 # Changelog
 
 ## [Unreleased]
+- maximum timeout increased to 300
+- add Support for 2022.3
 
 ## [0.13.14]
 - \#349 Extend support for Git Parameter. Thanks to @chrisxiao

--- a/build.gradle
+++ b/build.gradle
@@ -68,7 +68,7 @@ java {
 
 tasks.withType(JavaCompile) {
     options.encoding = 'UTF-8'
-    options.release.set(Integer.parseInt(readPropertyString("javaVersion")))
+    options.release.set(Integer.parseInt(readPropertyString("javaVersionTarget")))
 }
 tasks.wrapper {
     gradleVersion = readPropertyString("gradleVersion")

--- a/gradle.properties
+++ b/gradle.properties
@@ -31,6 +31,7 @@ platformPlugins = com.intellij.java
 
 # Java language level used to compile sources and to generate the files for - Java 11 is required since 2020.3
 javaVersion = 11
+javaVersionTarget = 11
 
 # Gradle Releases -> https://github.com/gradle/gradle/releases
 gradleVersion = 7.4.2

--- a/src/main/java/org/codinjutsu/tools/jenkins/logic/UrlBuilder.java
+++ b/src/main/java/org/codinjutsu/tools/jenkins/logic/UrlBuilder.java
@@ -16,19 +16,23 @@
 
 package org.codinjutsu.tools.jenkins.logic;
 
+import com.intellij.openapi.diagnostic.Logger;
 import com.intellij.openapi.project.Project;
-import org.apache.commons.httpclient.URIException;
-import org.apache.commons.httpclient.util.URIUtil;
+import lombok.SneakyThrows;
 import org.codinjutsu.tools.jenkins.JenkinsAppSettings;
+import org.codinjutsu.tools.jenkins.view.action.UploadPatchToJobAction;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
 import java.net.MalformedURLException;
 import java.net.URI;
+import java.net.URISyntaxException;
 import java.net.URL;
 import java.util.Optional;
 
 public class UrlBuilder {
+
+    private static final Logger LOG = Logger.getInstance(UploadPatchToJobAction.class.getName());
 
     private static final String API_JSON = "/api/json";
     private static final String BUILD = "/build";
@@ -65,7 +69,7 @@ public class UrlBuilder {
 
     public URL createRunJobUrl(String jobBuildUrl, JenkinsAppSettings configuration) {
         try {
-            String s = jobBuildUrl + URIUtil.encodePathQuery(String.format("%s?delay=%dsec", BUILD, configuration.getBuildDelay()));
+            String s = jobBuildUrl + encodePathQuery(String.format("%s?delay=%dsec", BUILD, configuration.getBuildDelay()));
             return new URL(s);
         } catch (Exception ex) {
             handleException(ex);
@@ -73,9 +77,40 @@ public class UrlBuilder {
         return null;
     }
 
+    @SneakyThrows
+    private static String encodePathQuery(String unescaped) {
+        final var queryIndex = unescaped.indexOf('?');
+        if (queryIndex == -1) {
+            return encodePath(unescaped);
+        } else {
+            var path = encodePath(unescaped.substring(0, queryIndex));
+            var query = unescaped.substring(queryIndex + 1);
+            try {
+                query = new URI(null, null, null, query, null)
+                        .toASCIIString().substring(1);  // remove leading ?
+            }
+            catch (URISyntaxException e) {
+                LOG.debug(e.getMessage(), e);
+            }
+            //return com.intellij.util.io.URLUtil.encodeQuery(query);
+            return path + '?' + query;
+        }
+    }
+
+    @SneakyThrows
+    private static String encodePath(String unescaped) {
+        try {
+            return new URI(null, null, unescaped, null, null).toASCIIString();
+        }
+        catch (URISyntaxException e) {
+            return unescaped;
+        }
+        //return com.intellij.util.io.URLUtil.encodePath(unescaped);
+    }
+
     public URL createStopBuildUrl(String buildUrl) {
         try {//http://jenkins.internal/job/it4em-it4em-DPD-GEOR-UAT-RO/27/stop
-            return new URL(buildUrl + URIUtil.encodePath("stop"));
+            return new URL(buildUrl + encodePath("stop"));
         } catch (Exception ex) {
             handleException(ex);
         }
@@ -84,7 +119,7 @@ public class UrlBuilder {
 
     public URL createJenkinsWorkspaceUrl(JenkinsAppSettings configuration) {
         try {
-            return new URL(URIUtil.encodePathQuery(configuration.getServerUrl() + API_JSON + TREE_PARAM + BASIC_JENKINS_INFO));
+            return new URL(encodePathQuery(configuration.getServerUrl() + API_JSON + TREE_PARAM + BASIC_JENKINS_INFO));
         } catch (Exception ex) {
             handleException(ex);
         }
@@ -97,7 +132,7 @@ public class UrlBuilder {
             basicViewInfo = CLOUDBEES_VIEW_INFO;
         }
         try {
-            return new URL(viewUrl + URIUtil.encodePathQuery(API_JSON + TREE_PARAM + basicViewInfo));
+            return new URL(viewUrl + encodePathQuery(API_JSON + TREE_PARAM + basicViewInfo));
         } catch (Exception ex) {
             handleException(ex);
         }
@@ -107,7 +142,7 @@ public class UrlBuilder {
 
     public URL createJobUrl(String jobUrl) {
         try {
-            return new URL(jobUrl + URIUtil.encodePathQuery(API_JSON + TREE_PARAM + BASIC_JOB_INFO));
+            return new URL(jobUrl + encodePathQuery(API_JSON + TREE_PARAM + BASIC_JOB_INFO));
         } catch (Exception ex) {
             handleException(ex);
         }
@@ -116,7 +151,7 @@ public class UrlBuilder {
 
     public URL createBuildUrl(String buildUrl) {
         try {
-            return new URL(buildUrl + URIUtil.encodePathQuery(API_JSON + TREE_PARAM + BASIC_BUILD_INFO));
+            return new URL(buildUrl + encodePathQuery(API_JSON + TREE_PARAM + BASIC_BUILD_INFO));
         } catch (Exception ex) {
             handleException(ex);
         }
@@ -125,7 +160,7 @@ public class UrlBuilder {
 
     public URL createBuildsUrl(String buildUrl) {
         try {
-            return new URL(buildUrl + URIUtil.encodePathQuery(API_JSON + TREE_PARAM + BASIC_BUILDS_INFO));
+            return new URL(buildUrl + encodePathQuery(API_JSON + TREE_PARAM + BASIC_BUILDS_INFO));
         } catch (Exception ex) {
             handleException(ex);
         }
@@ -143,7 +178,7 @@ public class UrlBuilder {
 
     public URL createAuthenticationUrl(String serverUrl) {
         try {
-            return new URL(serverUrl + URIUtil.encodePathQuery(API_JSON + TEST_CONNECTION_REQUEST));
+            return new URL(serverUrl + encodePathQuery(API_JSON + TEST_CONNECTION_REQUEST));
         } catch (Exception ex) {
             handleException(ex);
         }
@@ -162,14 +197,15 @@ public class UrlBuilder {
     private void handleException(Exception ex) {
         if (ex instanceof MalformedURLException) {
             throw new IllegalArgumentException("URL is malformed", ex);
-        } else if (ex instanceof URIException) {
-            throw new IllegalArgumentException(ERROR_DURING_URL_CREATION, ex);
         }
+//        else if (ex instanceof URIException) {
+//            throw new IllegalArgumentException(ERROR_DURING_URL_CREATION, ex);
+//        }
     }
 
     public URL createNestedJobUrl(String currentJobUrl) {
         try {
-            return new URL(currentJobUrl + URIUtil.encodePathQuery(API_JSON + TREE_PARAM + NESTED_JOBS_INFO));
+            return new URL(currentJobUrl + encodePathQuery(API_JSON + TREE_PARAM + NESTED_JOBS_INFO));
         } catch (Exception ex) {
             handleException(ex);
         }
@@ -180,7 +216,7 @@ public class UrlBuilder {
     @NotNull
     public URL createComputerUrl(String serverUrl) {
         try {
-            return new URL(serverUrl + URIUtil.encodePathQuery(COMPUTER + API_JSON + TREE_PARAM +
+            return new URL(serverUrl + encodePathQuery(COMPUTER + API_JSON + TREE_PARAM +
                     COMPUTER_INFO));
         } catch (Exception ex) {
             handleException(ex);
@@ -221,7 +257,7 @@ public class UrlBuilder {
 
     public URL createFillValueItemsUrl(String jobUrl, String className, String param) {
         try {
-            return new URL(jobUrl +  URIUtil.encodePathQuery(String.format(FILL_VALUE_ITEMS, className, param)));
+            return new URL(jobUrl +  encodePathQuery(String.format(FILL_VALUE_ITEMS, className, param)));
         } catch (Exception ex) {
             handleException(ex);
         }

--- a/src/main/java/org/codinjutsu/tools/jenkins/model/VirtualFilePart.java
+++ b/src/main/java/org/codinjutsu/tools/jenkins/model/VirtualFilePart.java
@@ -17,7 +17,6 @@
 package org.codinjutsu.tools.jenkins.model;
 
 import com.intellij.openapi.vfs.VirtualFile;
-import org.apache.commons.httpclient.methods.multipart.PartSource;
 
 import java.io.IOException;
 import java.io.InputStream;
@@ -27,25 +26,18 @@ import java.io.InputStream;
  *
  * @author Yuri Novitsky
  */
-public class VirtualFilePartSource implements PartSource {
+public class VirtualFilePart {
 
     private VirtualFile file;
 
-    public VirtualFilePartSource(VirtualFile file) {
+    public VirtualFilePart(VirtualFile file) {
         this.file = file;
     }
 
-    @Override
-    public long getLength() {
-        return file.getLength();
-    }
-
-    @Override
     public String getFileName() {
         return file.getName();
     }
 
-    @Override
     public InputStream createInputStream() throws IOException {
         return file.getInputStream();
     }

--- a/src/main/java/org/codinjutsu/tools/jenkins/security/BasicSecurityClient.java
+++ b/src/main/java/org/codinjutsu/tools/jenkins/security/BasicSecurityClient.java
@@ -16,20 +16,18 @@
 
 package org.codinjutsu.tools.jenkins.security;
 
-import org.apache.commons.httpclient.UsernamePasswordCredentials;
-import org.apache.commons.httpclient.auth.AuthScope;
-import org.apache.commons.httpclient.methods.PostMethod;
 import org.apache.commons.lang.StringUtils;
-import org.apache.http.HttpHeaders;
+import org.apache.http.HttpHost;
+import org.apache.http.HttpStatus;
+import org.apache.http.auth.UsernamePasswordCredentials;
+import org.apache.http.util.EntityUtils;
 import org.codinjutsu.tools.jenkins.exception.AuthenticationException;
 import org.codinjutsu.tools.jenkins.exception.ConfigurationException;
-import org.codinjutsu.tools.jenkins.util.IOUtils;
 import org.jetbrains.annotations.Nullable;
 
 import java.io.IOException;
-import java.io.InputStream;
-import java.net.HttpURLConnection;
 import java.net.URL;
+import java.util.Collections;
 
 
 class BasicSecurityClient extends DefaultSecurityClient {
@@ -53,28 +51,17 @@ class BasicSecurityClient extends DefaultSecurityClient {
     @Nullable
     private String doAuthentication(URL jenkinsUrl) throws AuthenticationException {
         if (StringUtils.isNotEmpty(username) && StringUtils.isNotEmpty(password)) {
-            httpClient.getState().setCredentials(
-                    new AuthScope(jenkinsUrl.getHost(), jenkinsUrl.getPort()),
-                    new UsernamePasswordCredentials(username, password));
+            final var targetHost = new HttpHost(jenkinsUrl.getHost(), jenkinsUrl.getPort(), jenkinsUrl.getProtocol());
+            final var credentials = new UsernamePasswordCredentials(username, password);
+            addAuthenticationPreemptive(targetHost, credentials);
         }
 
-        httpClient.getParams().setAuthenticationPreemptive(true);
-
-        PostMethod post = new PostMethod(jenkinsUrl.toString());
-        post.addRequestHeader(HttpHeaders.ACCEPT_LANGUAGE, "en-US,en;q=0.5");
-
+        final var post = createPost(jenkinsUrl.toString(), Collections.emptyList());
         try {
-            if (isCrumbDataSet()) {
-                post.addRequestHeader(jenkinsVersion.getCrumbName(), crumbData);
-            }
-
-            post.setDoAuthentication(true);
-            int responseCode = httpClient.executeMethod(post);
-            final String responseBody;
-            try(InputStream inputStream = post.getResponseBodyAsStream()) {
-                responseBody = IOUtils.toString(inputStream, post.getResponseCharSet());
-            }
-            if (responseCode != HttpURLConnection.HTTP_OK) {
+            final var response = executeHttp(post);
+            final var responseCode = response.getStatusLine().getStatusCode();
+            final var responseBody = EntityUtils.toString(response.getEntity());
+            if (responseCode != HttpStatus.SC_OK) {
                 checkResponse(responseCode, responseBody);
             }
             return responseBody;

--- a/src/main/java/org/codinjutsu/tools/jenkins/security/DefaultSecurityClient.java
+++ b/src/main/java/org/codinjutsu/tools/jenkins/security/DefaultSecurityClient.java
@@ -18,48 +18,94 @@ package org.codinjutsu.tools.jenkins.security;
 
 import com.github.cliftonlabs.json_simple.JsonObject;
 import com.github.cliftonlabs.json_simple.Jsoner;
-import org.apache.commons.httpclient.HttpClient;
-import org.apache.commons.httpclient.HttpMethodBase;
-import org.apache.commons.httpclient.MultiThreadedHttpConnectionManager;
-import org.apache.commons.httpclient.URIException;
-import org.apache.commons.httpclient.methods.PostMethod;
-import org.apache.commons.httpclient.methods.multipart.FilePart;
-import org.apache.commons.httpclient.methods.multipart.MultipartRequestEntity;
-import org.apache.commons.httpclient.methods.multipart.Part;
-import org.apache.commons.httpclient.methods.multipart.StringPart;
+import com.intellij.util.net.IdeHttpClientHelpers;
+import com.intellij.util.net.ssl.CertificateManager;
 import org.apache.commons.lang.StringUtils;
 import org.apache.http.HttpHeaders;
+import org.apache.http.HttpHost;
+import org.apache.http.HttpResponse;
+import org.apache.http.auth.AuthScope;
+import org.apache.http.auth.UsernamePasswordCredentials;
+import org.apache.http.client.AuthCache;
+import org.apache.http.client.CredentialsProvider;
+import org.apache.http.client.HttpClient;
+import org.apache.http.client.config.RequestConfig;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.client.methods.HttpUriRequest;
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.config.SocketConfig;
+import org.apache.http.entity.ContentType;
+import org.apache.http.entity.mime.MultipartEntityBuilder;
+import org.apache.http.impl.auth.BasicScheme;
+import org.apache.http.impl.client.BasicAuthCache;
+import org.apache.http.impl.client.BasicCredentialsProvider;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.impl.conn.PoolingHttpClientConnectionManager;
+import org.apache.http.util.EntityUtils;
 import org.codinjutsu.tools.jenkins.exception.AuthenticationException;
 import org.codinjutsu.tools.jenkins.exception.ConfigurationException;
+import org.codinjutsu.tools.jenkins.exception.JenkinsPluginRuntimeException;
 import org.codinjutsu.tools.jenkins.model.FileParameter;
 import org.codinjutsu.tools.jenkins.model.RequestData;
-import org.codinjutsu.tools.jenkins.model.VirtualFilePartSource;
-import org.codinjutsu.tools.jenkins.util.IOUtils;
+import org.codinjutsu.tools.jenkins.model.VirtualFilePart;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
+import javax.net.ssl.SSLContext;
 import java.io.IOException;
-import java.io.InputStream;
 import java.net.HttpURLConnection;
 import java.net.URL;
 import java.net.UnknownHostException;
+import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
-import java.util.ArrayList;
 import java.util.Collection;
 
 class DefaultSecurityClient implements SecurityClient {
 
     private static final String BAD_CRUMB_DATA = "No valid crumb was included in the request";
-    static final String CHARSET = StandardCharsets.UTF_8.name();
+    static final Charset CHARSET = StandardCharsets.UTF_8;
+    static final String CHARSET_NAME = CHARSET.name();
     protected final HttpClient httpClient;
-    private final int connectionTimout;
-    protected String crumbData;
+    protected @Deprecated String crumbData;
     protected JenkinsVersion jenkinsVersion = JenkinsVersion.VERSION_1;
+    private final CredentialsProvider credentialsProvider;
+    private final AuthCache authCache;
+    private final HttpClientContext httpClientContext;
 
     DefaultSecurityClient(String crumbData, int connectionTimout) {
-        this.httpClient = new HttpClient(new MultiThreadedHttpConnectionManager());
+        this(crumbData, connectionTimout, CertificateManager.getInstance().getSslContext(), true);
+    }
+
+    @SuppressWarnings("java:S2095")
+    DefaultSecurityClient(String crumbData, int connectionTimout, SSLContext sslContext, boolean useProxySettings) {
+        final var socketConfig = SocketConfig.custom()
+                .setSoTimeout(connectionTimout).build();
+
+        final var poolingHttpClientConnectionManager = new PoolingHttpClientConnectionManager();
+        poolingHttpClientConnectionManager.setDefaultSocketConfig(socketConfig);
+        final var requestConfig = RequestConfig.custom()
+                .setConnectTimeout(connectionTimout)
+                .setConnectionRequestTimeout(connectionTimout)
+                .setSocketTimeout(connectionTimout);
+        this.credentialsProvider = new BasicCredentialsProvider();
+        if (useProxySettings) {
+            IdeHttpClientHelpers.ApacheHttpClient4.setProxyIfEnabled(requestConfig);
+            IdeHttpClientHelpers.ApacheHttpClient4.setProxyCredentialsIfEnabled(credentialsProvider);
+        }
+
+        final var httpClientBuilder = HttpClients.custom()
+                .setSSLContext(sslContext)
+                .setDefaultSocketConfig(socketConfig)
+                .setDefaultRequestConfig(requestConfig.build())
+                .setDefaultCredentialsProvider(credentialsProvider)
+                .disableRedirectHandling();
+        this.httpClient = httpClientBuilder.build();
         this.crumbData = crumbData;
-        this.connectionTimout = connectionTimout;
+
+        this.authCache = new BasicAuthCache();
+        this.httpClientContext =  HttpClientContext.create();
+        this.httpClientContext.setCredentialsProvider(this.credentialsProvider);
+        this.httpClientContext.setAuthCache(authCache);
     }
 
     @Nullable
@@ -99,55 +145,62 @@ class DefaultSecurityClient implements SecurityClient {
      * </pre>
      */
     @NotNull
-    PostMethod createPost(String url, @NotNull Collection<RequestData> data) {
-        final PostMethod post = new PostMethod(url);
+    HttpPost createPost(String url, @NotNull Collection<RequestData> data) {
+        final var post = new HttpPost(url);
         if (!data.isEmpty()) {
-            final ArrayList<Part> parts = new ArrayList<>();
+            final var multipartEntity = MultipartEntityBuilder.create();
+
             data.stream().filter(FileParameter.class::isInstance)
                     .map(FileParameter.class::cast)
-                    .map(this::createFilePart)
-                    .forEach(parts::add);
-            final JsonObject parameterJson = new JsonObject();
+                    .forEach(fileParameter -> addMultipartBinaryBody(fileParameter, multipartEntity));
+
+            final var parameterJson = new JsonObject();
             parameterJson.put("parameter", data);
-            parts.add(new StringPart("json", Jsoner.serialize(parameterJson), DefaultSecurityClient.CHARSET));
-            post.setRequestEntity(new MultipartRequestEntity(parts.toArray(new Part[0]), post.getParams()));
+            multipartEntity.addTextBody("json", Jsoner.serialize(parameterJson), ContentType.APPLICATION_JSON)
+                    .setCharset(DefaultSecurityClient.CHARSET);
+            final var multipart = multipartEntity.build();
+            post.setEntity(multipart);
         }
 
+        if (isCrumbDataSet()) {
+            post.addHeader(jenkinsVersion.getCrumbName(), crumbData);
+        }
+        post.addHeader(HttpHeaders.ACCEPT_LANGUAGE, "en-US,en;q=0.5");
         return post;
     }
 
-    @NotNull
-    private FilePart createFilePart(FileParameter fileParameter) {
-        final String CONTENT_TYPE= null;
-        return new FilePart(fileParameter.getFileName(), new VirtualFilePartSource(fileParameter.getFile()),
-                CONTENT_TYPE, DefaultSecurityClient.CHARSET);
+    protected final void addAuthenticationPreemptive(HttpHost host, UsernamePasswordCredentials credentials) {
+        final var authScope = new AuthScope(host.getHostName(), host.getPort());
+        authCache.put(host, new BasicScheme());
+        credentialsProvider.setCredentials(authScope, credentials);
+    }
+
+    private void addMultipartBinaryBody(FileParameter fileParameter,
+                                        MultipartEntityBuilder multipartEntityBuilder) {
+        final var virtualFilePart = new VirtualFilePart(fileParameter.getFile());
+        try {
+            multipartEntityBuilder.addBinaryBody(fileParameter.getFileName(), virtualFilePart.createInputStream(),
+                    ContentType.APPLICATION_OCTET_STREAM, virtualFilePart.getFileName());
+        } catch (IOException e) {
+            throw new JenkinsPluginRuntimeException(e.getMessage(), e);
+        }
     }
 
     private void runMethod(String url, @NotNull Collection<RequestData> data, ResponseCollector responseCollector) {
-        final PostMethod post = createPost(url, data);
-
-        if (isCrumbDataSet()) {
-            post.addRequestHeader(jenkinsVersion.getCrumbName(), crumbData);
-        }
-        post.addRequestHeader(HttpHeaders.ACCEPT_LANGUAGE, "en-US,en;q=0.5");
-        post.getParams().setContentCharset(StandardCharsets.UTF_8.name());
+        final var post = createPost(url, data);
 
         try {
-            httpClient.getParams().setParameter("http.socket.timeout", connectionTimout);
-            httpClient.getParams().setParameter("http.connection.timeout", connectionTimout);
-
-            int statusCode = httpClient.executeMethod(post);
-            final String responseBody;
-            try (InputStream inputStream = post.getResponseBodyAsStream()) {
-                responseBody = IOUtils.toString(inputStream, post.getResponseCharSet());
-            }
+            final var response = executeHttp(post);
+            final var statusCode = response.getStatusLine().getStatusCode();
+            final var responseBody = EntityUtils.toString(response.getEntity());
             checkResponse(statusCode, responseBody);
 
             if (HttpURLConnection.HTTP_OK == statusCode) {
                 responseCollector.collect(statusCode, responseBody);
             }
+            // ToDo @mcmics enable redirection in httpclient again and try to remove following
             if (isRedirection(statusCode)) {
-                responseCollector.collect(statusCode, post.getResponseHeader("Location").getValue());
+                responseCollector.collect(statusCode, response.getLastHeader("Location").getValue());
             }
         } catch (UnknownHostException uhEx) {
             throw new ConfigurationException(String.format("Unknown server: %s", uhEx.getMessage()), uhEx);
@@ -157,6 +210,10 @@ class DefaultSecurityClient implements SecurityClient {
         } finally {
             post.releaseConnection();
         }
+    }
+
+    protected final HttpResponse executeHttp(HttpPost post) throws IOException {
+        return httpClient.execute(post, this.httpClientContext);
     }
 
     protected void checkResponse(int statusCode, String responseBody) throws AuthenticationException {
@@ -198,12 +255,8 @@ class DefaultSecurityClient implements SecurityClient {
         return url;
     }
 
-    protected final String createUrlForNotification(HttpMethodBase httpMethod) {
-        try {
-            return createUrlForNotification(httpMethod.getURI().toString());
-        } catch (URIException e) {
-            throw new ConfigurationException(String.format("Invalid URI: %s", e.getMessage()), e);
-        }
+    protected final String createUrlForNotification(HttpUriRequest httpMethod) {
+        return createUrlForNotification(httpMethod.getURI().toString());
     }
 
     private static class ResponseCollector {

--- a/src/test/java/org/codinjutsu/tools/jenkins/logic/UrlBuilderTest.java
+++ b/src/test/java/org/codinjutsu/tools/jenkins/logic/UrlBuilderTest.java
@@ -21,8 +21,6 @@ import org.junit.Before;
 import org.junit.Test;
 
 import java.net.URL;
-import java.util.LinkedHashMap;
-import java.util.Map;
 
 import static org.assertj.core.api.Assertions.*;
 
@@ -45,25 +43,26 @@ public class UrlBuilderTest {
         configuration.setServerUrl("http://localhost:8080/jenkins");
 
         URL url = urlBuilder.createJenkinsWorkspaceUrl(configuration);
-        assertThat(url).hasToString("http://localhost:8080/jenkins/api/json?tree=url,description,nodeName,nodeDescription,primaryView%5Bname,url%5D,views%5Bname,url,views%5Bname,url%5D%5D");
+        assertThat(url).hasToString("http://localhost:8080/jenkins/api/json?tree=url,description,nodeName,nodeDescription,primaryView[name,url],views[name,url,views[name,url]]");
     }
 
     @Test
     public void createViewUrlForClassicPlateform() {
         URL url = urlBuilder.createViewUrl(JenkinsPlateform.CLASSIC, "http://localhost:8080/jenkins/My%20View");
-        assertThat(url).hasToString("http://localhost:8080/jenkins/My%20View/api/json?tree=name,url,jobs%5Bname,fullName,displayName,fullDisplayName,jobs,url,color,buildable,inQueue,healthReport%5Bdescription,iconUrl%5D,lastBuild%5Burl,id,building,result,number,displayName,fullDisplayName,timestamp,duration,actions%5Bparameters%5Bname,value%5D%5D%5D,lastFailedBuild%5Burl%5D,lastSuccessfulBuild%5Burl%5D,property%5BparameterDefinitions%5Bname,type,defaultParameterValue%5Bvalue%5D,description,choices%5D%5D%5D");
+        assertThat(url).hasToString("http://localhost:8080/jenkins/My%20View/api/json?tree=name,url,jobs[name,fullName,displayName,fullDisplayName,jobs,url,color,buildable,inQueue,healthReport[description,iconUrl],lastBuild[url,id,building,result,number,displayName,fullDisplayName,timestamp,duration,actions[parameters[name,value]]],lastFailedBuild[url],lastSuccessfulBuild[url],property[parameterDefinitions[name,type,defaultParameterValue[value],description,choices]]]");
     }
 
     @Test
     public void createJobJSONUrl() {
         URL url = urlBuilder.createJobUrl("http://localhost:8080/jenkins/my%20Job");
-        assertThat(url).hasToString("http://localhost:8080/jenkins/my%20Job/api/json?tree=name,fullName,displayName,fullDisplayName,jobs,url,color,buildable,inQueue,healthReport%5Bdescription,iconUrl%5D,lastBuild%5Burl,id,building,result,number,displayName,fullDisplayName,timestamp,duration,actions%5Bparameters%5Bname,value%5D%5D%5D,lastFailedBuild%5Burl%5D,lastSuccessfulBuild%5Burl%5D,property%5BparameterDefinitions%5Bname,type,defaultParameterValue%5Bvalue%5D,description,choices%5D%5D");
+        assertThat(url).hasToString("http://localhost:8080/jenkins/my%20Job/api/json?tree=name,fullName,displayName,fullDisplayName,jobs,url,color,buildable,inQueue,healthReport[description,iconUrl],lastBuild[url,id,building,result,number,displayName,fullDisplayName,timestamp,duration,actions[parameters[name,value]]],lastFailedBuild[url],lastSuccessfulBuild[url],property[parameterDefinitions[name,type,defaultParameterValue[value],description,choices]]");
     }
 
     @Test
     public void createViewUrlForCloudbeesPlateform() {
         URL url = urlBuilder.createViewUrl(JenkinsPlateform.CLOUDBEES, "http://localhost:8080/jenkins/My%20View");
-        assertThat(url).hasToString("http://localhost:8080/jenkins/My%20View/api/json?tree=name,url,views%5Bjobs%5Bname,fullName,displayName,fullDisplayName,jobs,url,color,buildable,inQueue,healthReport%5Bdescription,iconUrl%5D,lastBuild%5Burl,id,building,result,number,displayName,fullDisplayName,timestamp,duration,actions%5Bparameters%5Bname,value%5D%5D%5D,lastFailedBuild%5Burl%5D,lastSuccessfulBuild%5Burl%5D,property%5BparameterDefinitions%5Bname,type,defaultParameterValue%5Bvalue%5D,description,choices%5D%5D%5D%5D");
+        assertThat(url).hasToString("http://localhost:8080/jenkins/My%20View/api/json?tree=name,url,views[jobs[name,fullName,displayName,fullDisplayName,jobs,url,color,buildable,inQueue,healthReport[description,iconUrl],lastBuild[url,id,building,result,number,displayName,fullDisplayName,timestamp,duration,actions[parameters[name,value]]],lastFailedBuild[url],lastSuccessfulBuild[url],property[parameterDefinitions[name,type,defaultParameterValue[value],description,choices]]]]");
+        assertThat(url.toString()).contains("/api/json?tree=name,url,views[jobs[name,fullName,displayName,fullDisplayName,jobs,url,color,buildable,inQueue,healthReport[description,iconUrl],lastBuild[url,id,building,result,number,displayName,fullDisplayName,timestamp,duration,actions[parameters[name,value]]],lastFailedBuild[url],lastSuccessfulBuild[url],property[parameterDefinitions[name,type,defaultParameterValue[value],description,choices]]]]");
     }
 
     @Test
@@ -75,13 +74,13 @@ public class UrlBuilderTest {
     @Test
     public void createAuthenticationJSONUrl() {
         URL url = urlBuilder.createAuthenticationUrl("http://localhost:8080/jenkins");
-        assertThat(url).hasToString("http://localhost:8080/jenkins/api/json?tree=nodeName,url,description,primaryView%5Bname,url%5D");
+        assertThat(url).hasToString("http://localhost:8080/jenkins/api/json?tree=nodeName,url,description,primaryView[name,url]");
     }
 
     @Test
     public void createComputerJSONUrl() {
         URL url = urlBuilder.createComputerUrl("http://localhost:8080/jenkins");
-        assertThat(url).hasToString("http://localhost:8080/jenkins/computer/api/json?tree=computer%5BdisplayName,description,offline,assignedLabels%5Bname%5D%5D");
+        assertThat(url).hasToString("http://localhost:8080/jenkins/computer/api/json?tree=computer[displayName,description,offline,assignedLabels[name]]");
     }
 
     @Before

--- a/src/test/java/org/codinjutsu/tools/jenkins/security/DefaultSecurityClientTest.java
+++ b/src/test/java/org/codinjutsu/tools/jenkins/security/DefaultSecurityClientTest.java
@@ -1,109 +1,134 @@
 package org.codinjutsu.tools.jenkins.security;
 
-import com.github.cliftonlabs.json_simple.Jsoner;
 import com.intellij.mock.MockVirtualFile;
 import com.intellij.openapi.vfs.VirtualFile;
-import org.apache.commons.httpclient.methods.PostMethod;
-import org.apache.commons.httpclient.methods.multipart.FilePart;
-import org.apache.commons.httpclient.methods.multipart.MultipartRequestEntity;
-import org.apache.commons.httpclient.methods.multipart.Part;
-import org.apache.commons.httpclient.methods.multipart.StringPart;
+import lombok.SneakyThrows;
+import org.apache.http.HttpEntity;
+import org.apache.http.entity.mime.FormBodyPart;
+import org.apache.http.entity.mime.content.InputStreamBody;
+import org.apache.http.entity.mime.content.StringBody;
 import org.assertj.core.api.Assertions;
 import org.assertj.core.util.Lists;
 import org.codinjutsu.tools.jenkins.model.FileParameter;
 import org.codinjutsu.tools.jenkins.model.RequestData;
 import org.codinjutsu.tools.jenkins.model.StringParameter;
+import org.jetbrains.annotations.NotNull;
 import org.junit.Test;
 import org.powermock.reflect.Whitebox;
 
+import javax.net.ssl.SSLContext;
+import java.io.ByteArrayInputStream;
 import java.io.IOException;
-import java.lang.reflect.Field;
+import java.io.InputStream;
 import java.nio.charset.StandardCharsets;
+import java.security.NoSuchAlgorithmException;
 import java.util.Collections;
 import java.util.List;
-import java.util.Set;
-
-import static org.junit.Assert.*;
 
 public class DefaultSecurityClientTest {
 
-    private final DefaultSecurityClient securityClient = new DefaultSecurityClient("Crumb", 10);
+    private final DefaultSecurityClient securityClient;
+
+    public DefaultSecurityClientTest() throws NoSuchAlgorithmException {
+        securityClient = new DefaultSecurityClient("Crumb", 10, SSLContext.getDefault(), false);
+    }
+
 
     @Test
     public void createPostForEmptyData() {
-        final PostMethod post = securityClient.createPost("http://example.org", Collections.emptySet());
-        Assertions.assertThat(post.getRequestEntity()).isNull();
+        final var post = securityClient.createPost("http://example.org", Collections.emptySet());
+        Assertions.assertThat(post.getEntity()).isNull();
     }
 
     @Test
     public void createPostWithOneFileAndOneStringParameter() throws IOException {
-        final VirtualFile virtualFile = new MockVirtualFile("sampleFile.md");
-        final PostMethod post = securityClient.createPost("http://example.org", Lists.list(new StringParameter("test", "Jenkins"),
+        final VirtualFile virtualFile = new MockVirtualFileWithInputStream("sampleFile.md");
+        final var post = securityClient.createPost("http://example.org", Lists.list(new StringParameter("test", "Jenkins"),
                 new FileParameter("fileParam", virtualFile)));
-        Assertions.assertThat(post.getRequestEntity()).isNotNull();
-        Assertions.assertThat(post.getRequestEntity()).isInstanceOf(MultipartRequestEntity.class);
-        final MultipartRequestEntity multipartRequestEntity = (MultipartRequestEntity) post.getRequestEntity();
-        final Part[] parts = Whitebox.getInternalState(multipartRequestEntity, Part[].class);
+        Assertions.assertThat(post.getEntity()).isNotNull();
+        final List<FormBodyPart> parts = assertMultipartFormEntity(post.getEntity());
         Assertions.assertThat(parts).hasSize(2);
-        Assertions.assertThat(parts[0]).isInstanceOf(FilePart.class);
-        Assertions.assertThat(parts[0].getName()).isEqualTo("sampleFile.md");
-        Assertions.assertThat(parts[1]).isInstanceOf(StringPart.class);
+        final FormBodyPart firstPart = parts.get(0);
+        final FormBodyPart secondPart = parts.get(1);
+        Assertions.assertThat(firstPart.getName()).isEqualTo("sampleFile.md");
+        Assertions.assertThat(secondPart.getName()).isEqualTo("json");
+        Assertions.assertThat(firstPart.getBody()).isInstanceOf(InputStreamBody.class);
+        Assertions.assertThat(secondPart.getBody()).isInstanceOf(StringBody.class);
         final String expectedJson = "{\"parameter\":[{\"name\":\"fileParam\",\"file\":\"sampleFile.md\"},{\"name\":\"test\",\"value\":\"Jenkins\"}]}";
-        final StringPart expected = new StringPart("json", expectedJson, DefaultSecurityClient.CHARSET);
-        Assertions.assertThat(parts[1].length()).isEqualTo(expected.length());
+        Assertions.assertThat(secondPart.getBody().getContentLength()).isEqualTo(expectedJson.length());
     }
 
     @Test
     public void createPostWithFileNameProvider() throws IOException {
-        final VirtualFile virtualFile = new MockVirtualFile("sampleFile.md");
-        final PostMethod post = securityClient.createPost("http://example.org", Lists.list(new StringParameter("test", "Jenkins"),
+        final VirtualFile virtualFile = new MockVirtualFileWithInputStream("sampleFile.md");
+        final var post = securityClient.createPost("http://example.org", Lists.list(new StringParameter("test", "Jenkins"),
                 new FileParameter("fileParam", virtualFile, () -> "file0")));
-        Assertions.assertThat(post.getRequestEntity()).isNotNull();
-        Assertions.assertThat(post.getRequestEntity()).isInstanceOf(MultipartRequestEntity.class);
-        final MultipartRequestEntity multipartRequestEntity = (MultipartRequestEntity) post.getRequestEntity();
-        final Part[] parts = Whitebox.getInternalState(multipartRequestEntity, Part[].class);
+        Assertions.assertThat(post.getEntity()).isNotNull();
+        final var parts = assertMultipartFormEntity(post.getEntity());
         Assertions.assertThat(parts).hasSize(2);
-        Assertions.assertThat(parts[0]).isInstanceOf(FilePart.class);
-        Assertions.assertThat(parts[0].getName()).isEqualTo("file0");
-        Assertions.assertThat(parts[1]).isInstanceOf(StringPart.class);
+        final FormBodyPart firstPart = parts.get(0);
+        final FormBodyPart secondPart = parts.get(1);
+        Assertions.assertThat(firstPart.getName()).isEqualTo("file0");
+        Assertions.assertThat(secondPart.getName()).isEqualTo("json");
+        Assertions.assertThat(firstPart.getBody()).isInstanceOf(InputStreamBody.class);
+        Assertions.assertThat(secondPart.getBody()).isInstanceOf(StringBody.class);
         final String expectedJson = "{\"parameter\":[{\"name\":\"fileParam\",\"file\":\"file0\"},{\"name\":\"test\",\"value\":\"Jenkins\"}]}";
-        final StringPart expected = new StringPart("json", expectedJson, DefaultSecurityClient.CHARSET);
-        Assertions.assertThat(parts[1].length()).isEqualTo(expected.length());
+        Assertions.assertThat(secondPart.getBody().getContentLength()).isEqualTo(expectedJson.length());
     }
 
     @Test
     public void createPostWithTwoStringParameter() throws IOException {
-        final List<RequestData> requestData =Lists.list(new StringParameter("test", "Jenkins"),
+        final List<RequestData> requestData = Lists.list(new StringParameter("test", "Jenkins"),
                 new StringParameter("second", "more"));
-        final PostMethod post = securityClient.createPost("http://example.org", requestData);
-        Assertions.assertThat(post.getRequestEntity()).isNotNull();
-        Assertions.assertThat(post.getRequestEntity()).isInstanceOf(MultipartRequestEntity.class);
-        final MultipartRequestEntity multipartRequestEntity = (MultipartRequestEntity) post.getRequestEntity();
-        final Part[] parts = Whitebox.getInternalState(multipartRequestEntity, Part[].class);
+        final var post = securityClient.createPost("http://example.org", requestData);
+        Assertions.assertThat(post.getEntity()).isNotNull();
+        final var parts = assertMultipartFormEntity(post.getEntity());
         Assertions.assertThat(parts).hasSize(1);
-        Assertions.assertThat(parts[0]).isInstanceOf(StringPart.class);
-        Assertions.assertThat(parts[0].getName()).isEqualTo("json");
+        final FormBodyPart firstPart = parts.get(0);
+        Assertions.assertThat(firstPart.getName()).isEqualTo("json");
+        Assertions.assertThat(firstPart.getBody()).isInstanceOf(StringBody.class);
         final String expectedJson = "{\"parameter\":[{\"name\":\"test\",\"value\":\"Jenkins\"},{\"name\":\"second\",\"value\":\"more\"}]}";
-        final StringPart expected = new StringPart("json", expectedJson, DefaultSecurityClient.CHARSET);
-
-        Assertions.assertThat(parts[0].length()).isEqualTo(expected.length());
+        Assertions.assertThat(firstPart.getBody().getContentLength()).isEqualTo(expectedJson.length());
     }
 
     @Test
     public void createPostWithParameterIsEncodedAsUTF8() throws IOException {
-        final List<RequestData> requestData =Lists.list(new StringParameter("first", "İstanbul"),
+        final List<RequestData> requestData = Lists.list(new StringParameter("first", "İstanbul"),
                 new StringParameter("second", "İÖÇŞĞ"));
-        final PostMethod post = securityClient.createPost("http://example.org", requestData);
-        Assertions.assertThat(post.getRequestEntity()).isNotNull();
-        Assertions.assertThat(post.getRequestEntity()).isInstanceOf(MultipartRequestEntity.class);
-        final MultipartRequestEntity multipartRequestEntity = (MultipartRequestEntity) post.getRequestEntity();
-        final Part[] parts = Whitebox.getInternalState(multipartRequestEntity, Part[].class);
-        Assertions.assertThat(parts).hasSize(1);
-        Assertions.assertThat(parts[0]).isInstanceOf(StringPart.class);
-        Assertions.assertThat(parts[0].getName()).isEqualTo("json");
-        final String expectedJson = "{\"parameter\":[{\"name\":\"first\",\"value\":\"İstanbul\"},{\"name\":\"second\",\"value\":\"İÖÇŞĞ\"}]}";
-        final StringPart expected = new StringPart("json", expectedJson, DefaultSecurityClient.CHARSET);
+        final var post = securityClient.createPost("http://example.org", requestData);
+        Assertions.assertThat(post.getEntity()).isNotNull();
 
-        Assertions.assertThat(parts[0].length()).isEqualTo(expected.length());
+        final var parts = assertMultipartFormEntity(post.getEntity());
+        Assertions.assertThat(parts).hasSize(1);
+        final FormBodyPart firstPart = parts.get(0);
+        Assertions.assertThat(firstPart.getName()).isEqualTo("json");
+        Assertions.assertThat(firstPart.getBody()).isInstanceOf(StringBody.class);
+        final String expectedJson = "{\"parameter\":[{\"name\":\"first\",\"value\":\"İstanbul\"},{\"name\":\"second\",\"value\":\"İÖÇŞĞ\"}]}";
+        Assertions.assertThat(firstPart.getBody().getContentLength())
+                .isEqualTo(expectedJson.getBytes(StandardCharsets.UTF_8).length);
+    }
+
+    /**
+     * @see org.apache.http.entity.mime.MultipartFormEntity
+     * @see org.apache.http.entity.mime.AbstractMultipartForm
+     * @see org.apache.http.entity.mime.AbstractMultipartForm#getBodyParts()
+     */
+    @SneakyThrows
+    private List<FormBodyPart> assertMultipartFormEntity(HttpEntity httpEntity) {
+        Assertions.assertThat(httpEntity.getClass().getName()).isEqualTo("org.apache.http.entity.mime.MultipartFormEntity");
+        final var abstractMultipartForm = Whitebox.getInternalState(httpEntity, "multipart");
+        return Whitebox.invokeMethod(abstractMultipartForm, "getBodyParts");
+    }
+
+    private class MockVirtualFileWithInputStream extends MockVirtualFile {
+
+        public MockVirtualFileWithInputStream(String name) {
+            super(name);
+        }
+
+        @Override
+        public @NotNull InputStream getInputStream() {
+            return new ByteArrayInputStream(getName().getBytes(StandardCharsets.UTF_8));
+        }
     }
 }

